### PR TITLE
More time-out checks and a k-limit check for expression refinement.

### PR DESCRIPTION
**Pull request by hermanventer**
## Description

Some funky functions take a long time to do things that were atomic w.r.t. time out checks. Made time-outs more granular. Also added a check that prevents very large expressions from being refined, thus trading precision for analysis speed.

## Type of change

- [x] Performance bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API change with a documentation update
- [ ] Additional test coverage

## How Has This Been Tested?
cargo test; ./validate.sh

### CODE REVIEW

File: checker/src/visitors.rs [hunk 1]
```diff
@@ -1596,6 +1596,10 @@ impl<'a, 'b: 'a, 'tcx: 'b, E> MirVisitor<'a, 'b, 'tcx, E> {
                     break;
                 }
             }
+            let elapsed_time_in_seconds = self.start_instant.elapsed().as_secs();
+            if elapsed_time_in_seconds >= k_limits::MAX_ANALYSIS_TIME_FOR_BODY {
+                return;
```
Comment (zlandau) [inline checker/src/visitors.rs]: Do you keep stats on any of these types of timeouts? Isn't it hard to track down problems when anything can just silently give up?

File: checker/src/visitors.rs [hunk 2]
```diff
@@ -1960,6 +1972,10 @@ impl<'a, 'b: 'a, 'tcx: 'b, E> MirVisitor<'a, 'b, 'tcx, E> {
             .iter()
             .filter(|(p, _)| p.is_rooted_by(&rpath))
         {
+            let elapsed_time_in_seconds = self.start_instant.elapsed().as_secs();
```
Comment (zlandau) [inline checker/src/visitors.rs]: This is repeated a lot, isn't there somewhere you can put a reached_max_analysis_time? I guess it's cheating to put it into k_limits itself.

File: checker/src/visitors.rs (hunk 1 above)
Comment (hermanventer) [reply checker/src/visitors.rs]: Time-outs surface as WARN level entries in log output. I am painfully aware of any time-out.

File: checker/src/visitors.rs (hunk 2 above)
Comment (hermanventer) [reply checker/src/visitors.rs]: It is bothering me a bit as well, but not as much as other things. Perhaps I'll come back to this in the future.

**Pull request merged**